lib/tasks/email.rake

```diff
@@ -1,107 +0,0 @@
-    end
-    config = Rails.application.credentials.email
-    
-    reconnectSleep = 1
-    
-    logger.info("Logging in to mailbox #{config[:email]}")
-    while true
-      begin
-        imap = Net::IMAP.new(config[:host], port: config[:port], ssl: config[:ssl])
-        imap.capable?(:IMAP4rev1) or raise "Not an IMAP4rev1 server"
-        if imap.auth_capable?("XOAUTH2") && !Rails.application.credentials.email[:oauth].nil?
-          oauth_client = OAuth2::Client.new(config[:oauth][:client_id], config[:oauth][:client_secret], {site: config[:oauth][:site], authorize_url: config[:oauth][:authorize_url], token_url: config[:oauth][:token_url]})
-          access_token = OAuth2::AccessToken.from_hash(oauth_client, refresh_token: config[:oauth][:refresh_token]).refresh!
-          imap.authenticate('XOAUTH2', config[:email], access_token.token)
-        elsif imap.auth_capable?("PLAIN")
-          imap.authenticate("PLAIN", config[:email], config[:password])
-        # Should not use deprecated LOGIN method
-        # elsif !imap.capability?("LOGINDISABLED")
-          # imap.login(config[:email], config[:password])
-        else
-          raise "No acceptable authentication mechanisms"
-        end
-      rescue Net::IMAP::NoResponseError, SocketError, Faraday::ConnectionFailed => error
-        logger.info("Could not authenticate for #{config[:email]}, trying again in #{reconnectSleep} #{"second".pluralize(reconnectSleep)}, error: #{error.message}")
-        sleep reconnectSleep
-        reconnectSleep += 1
-        
-        next
-      end
-      
-      reconnectSleep = 1
-      
-      begin
-        imap.select(config[:name])
-        
-        while true
-          logger.info("Pulling emails for #{config[:email]}")
-          query = ["BEFORE", Net::IMAP.format_date(Time.now + 1.day)]
-          latest = Email.order("timestamp DESC").first
-          query = ["SINCE", Net::IMAP.format_date(latest.timestamp)] if latest
-          
-          ids = imap.search(query)
-          imap.fetch(ids, "BODY.PEEK[]").each do |msg|
-            mail = Mail.new(msg.attr["BODY[]"])
-            
-            unless Email.where(message_id: mail.message_id).exists?
-              begin
-                unless Email.create_from_mail(mail)
-                  logger.error("Could not pull message #{mail.message_id}")
-                end
-              rescue Exception => e
-                logger.error("Exception while loading message #{mail.message_id}: " + e.to_s + "\n" + e.backtrace.join("\n"))
-              end
-            end
-          end
-          
-          logger.info("Idling for #{config[:email]}")
-          
-          waiting = Thread.start do
-            sleep(20.minutes)
-            
-            imap.idle_done
-          end
-          
-          imap.idle do |response|
-            if response.respond_to?(:name) && response.name == 'EXISTS'
-              waiting.kill
-              imap.idle_done
-            end
-          end
```

Comment:
On some level I'd rather have more frequent short-lived jobs than long-running ones. Sidekiq at least can run multiple threads, but it still hogs one of them. However, with Docker scaling plus threading, I don't think it will matter to us, so I'm happy to re-roll the idling.

Another concern of mine is that sounds like if we immediately reschedule, it would be also possible to end up in a fail-loop and get rate-limited. I think if we do idle, we should just resume at the next scheduled time.

Reply:
I think we could experiment with a shorter idle delay--maybe 60 seconds to 5 minutes. That way a job never runs longer than 5 minutes. If it fails, it seems there is a `wait` parameter on the `retry_on` method so we can have it wait 10 minutes if it fails. That should cover the looping-rate-limit issue. It could then re-queue itself or rely on another scheduler (although we would need to scheduler to not run if a retry is currently waiting).

https://edgeapi.rubyonrails.org/classes/ActiveJob/Exceptions/ClassMethods.html#method-i-retry_on

So 1) initial job starts, 2) job ends and it goes into idle (either launches an idle job or just continues current job to idle), 3) idle job ends and re-queue or scheduler starts another.

I'm not really worried about thread hogging. Even though the task is running, this is blocking I/O so it won't consume anything else on the system. We can just treat our scaling as `n + 1` where 1 is a worker dedicated to email. But also Sidekiq might recognize the blocking I/O and not lock up that worker (I have not tested this).